### PR TITLE
Infra/csp headers

### LINKED ISSUE #64: CSP Headers
## Full User Story
The frontend domain(s) should respond with an appropriate set of Content Security Policy headers. These headers instruct the user's browser to not allow unexpected content to load.

### Dependencies
- The domains for each environment need to be known OR we need to come back to this and update when known

### Acceptance Criteria
- A passing score on `https://securityheaders.com/
- IA completes code review
- QA: regression test

### PR DESCRIPTION
**Pull request by jsandoval81**
### Requirements List
- **New CSP Lambda**
    - `/backend/compact-connect/lambdas/cloudfront-csp`
    - Follow README setup
- **Browser frontend**
    - `/webroot`
    - `yarn install --ignore-engines`

### Description List
- Filled out the `cloudfront-csp` lambda function to respond with security headers, including CSP
- Updated the browser frontend app to address any CSP items
    - Updated some SVGs to remove inline styles
    - Updated the way translations are handled & built with `vue-i18n`
- Updated the deployment pipeline with updated domains

### Draft Todo
- [x] Confirm updated IA deployment pipelines (if possible) _**(IA)**_
- [x] Remove any feature branches from pipeline files (if possible) _**(IA)**_

_Note: This is one of the last (or the last?) features cut from the "old" way of managing trunks - so it has a messy commit history. Moving forward, we're using a cleaner process for managing trunks across repos._

Closes #64

### CODE REVIEW

**Comment by jsandoval81** [comment]
@jlkravitz This is ready for review.

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] ~stand up the site locally~
- [ ] review static code analysis results
- [ ] run a security audit of dependencies (e.g. `npm audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)
- [ ] examine OWASP ZAP output to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

File: backend/compact-connect/tests/unit/test_app.py [hunk 1]
```diff
@@ -95,6 +95,7 @@ def test_synth_sandbox_no_domain(self):
         self._check_no_annotations(app.sandbox_stage.persistent_stack)
         self._check_no_annotations(app.sandbox_stage.ui_stack)
         self._check_no_annotations(app.sandbox_stage.api_stack)
+        self._check_no_annotations(app.sandbox_stage.ingest_stack)
```
Comment (jlkravitz) [inline backend/compact-connect/tests/unit/test_app.py:98]: This related to this PR or just caught something from a past sprint?

File: webroot/vue.config.js [hunk 2]
```diff
@@ -296,6 +306,9 @@ module.exports = {
         plugins: [
             faviconsPlugin,
             stylelintPlugin,
+            // The runtime-only version of vue-i18n conflicts with test-runner, but is required for built server versions due to CSP.
```
Comment (jlkravitz) [inline webroot/vue.config.js:309]: What is test-runner?

File: webroot/src/models/State/State.model.ts [hunk 3]
```diff
@@ -38,7 +38,17 @@ export class State implements InterfaceStateCreate {
...
+        let states = this.$tm('common.states') || [];
+
+        if (typeof states[0]?.abbrev === 'function') {
+            const normalize = ([value]) => value;
+
+            states = states.map((st) => ({
+                abbrev: st.abbrev({ normalize }),
+                full: st.full({ normalize }),
+            }));
+        }
```
Comment (jlkravitz) [inline webroot/src/models/State/State.model.ts:50]: What's going on here? (I understand step-by-step what's happening, but wondering why it needs to be done, and why `abbrev` might be something other than a string.)

Same question on `StateUpload.ts` but I assume the answer is the same/similar.

File: backend/compact-connect/lambdas/cloudfront-csp/test/config/index.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+//  CompactConnect
+//
+//  Created by InspiringApps on 7/22/2024.
+//
+
+const chaiMatchPattern = require('chai-match-pattern');
+const chai = require('chai').use(chaiMatchPattern);
+const chalk = require('chalk');
+const lambdaLocal = require('lambda-local');
+const path = require('path');
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/config/index.js:12]: What's with the CommonJS `require` imports? Something to do with lambda requirements or environment?

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    ].join(' ');
+    const cspScriptSrcElem = [
+        '\'self\'',
+    ].join(' ');
+    const cspScriptSrcAttr = [
+        '\'self\'',
+    ].join(' ');
+    const cspWorkerSrc = [
+        '\'self\'',
+    ].join(' ');
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:69]: I assume your goal here is to let this be extendable going forward? It's definitely less readable to me than just e.g., `cspDefaultSrc = "'none'";` but perhaps worth it. Mind adding a comment at the top just indicating why it's written like this?

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+    const cspFontSrc = [
+        '\'self\'',
+        'https://fonts.gstatic.com',
+    ].join(' ');
+    const cspImgSrc = [
+        '\'self\'',
+        'data:',
+        dataApiUrl,
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:88]: What's this doing?

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+        '\'none\'',
+    ].join(' ');
+    const cspFormActionSrc = [
+        '\'none\'',
+    ].join(' ');
+    const cspConnectSrc = [
+        '\'self\'',
+        dataApiUrl,
+        s3Url,
+        'https://cognito-idp.us-east-1.amazonaws.com'
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:110]: What's this URL? Perhaps hoist up to top of function and label?

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+        `style-src-attr ${cspStyleSrcAttr};`,
+        `font-src ${cspFontSrc};`,
+        `img-src ${cspImgSrc};`,
+        `media-src ${cspMediaSrc};`,
+        `frame-src ${cspFrameSrc};`,
+        `frame-ancestors ${cspFrameAncestorsSrc};`,
+        `object-src ${cspObjectSrc};`,
+        `form-action ${cspFormActionSrc};`,
+        `connect-src ${cspConnectSrc};`,
+    ].join(' ')}`;
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:131]: 🎊

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+                headers: {},
+            };
+            const config = lambdaConfig({
+                lambdaPath: `index.js`,
+                request,
+                response,
+            });
+            const result = await runLambda(config);
+
+            // console.log(JSON.stringify(result, null, 2)); // @DEBUG
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:158]: remove debug code

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+
+            expect(result.headers['strict-transport-security'][0].key).to.equal('Strict-Transport-Security');
+            expect(result.headers['strict-transport-security'][0].value).to.equal('max-age=31536000; includeSubdomains; preload');
+            expect(result.headers['x-content-type-options'][0].key).to.equal('X-Content-Type-Options');
+            expect(result.headers['x-content-type-options'][0].value).to.equal('nosniff');
+            expect(result.headers['x-frame-options'][0].key).to.equal('X-Frame-Options');
+            expect(result.headers['x-frame-options'][0].value).to.equal('DENY');
+            expect(result.headers['x-xss-protection'][0].key).to.equal('X-Xss-Protection');
+            expect(result.headers['x-xss-protection'][0].value).to.equal('1; mode=block');
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:167]: This looks to be nonstandard. If you think we need it, mind adding a comment as to why, and perhaps point out why (or how) the vulnerability it describes is not an issue for us?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+            expect(result.headers['referrer-policy'][0].key).to.equal('Referrer-Policy');
+            expect(result.headers['referrer-policy'][0].value).to.equal('strict-origin-when-cross-origin');
+            expect(result.headers['content-security-policy'][0].key).to.equal('Content-Security-Policy');
+            expect(result.headers['content-security-policy'][0].value).to.equal(buildCspHeaders(environment));
+            expect(result.headers.server[0].key).to.equal('Server');
+            expect(result.headers.server[0].value).to.equal('CompactConnect');
+        });
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:318]: I know this was a last discussion on last PR... but I think this might cross the line of being explicit on tests. I _expect_ that these tests are all identical, but I really can't tell easily. Could we refactor? It seems the only thing changing is the environment variable.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+        it('should successfully return the security headers for csg prod when lambda event is missing the request domain', async () => {
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:355]: Thanks for testing these cases!

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 13]
```diff
@@ -1,5 +1,337 @@
...
+/**
+ * Configuration of supported domains per environment.
+ * @type {object}
+ */
+const environments = {
+    csg: {
+        prod: {
+            webFrontend: `app.compactconnect.org`,
+            dataApi: `api.compactconnect.org`,
+            s3Upload: ``, // @TODO: Waiting for environment configuration.
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:20]: WIll this be addressed for this PR, or is it a backlog item? If the latter, can we make sure it's being tracked elsewhere besides the code?

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 14]
```diff
@@ -1,5 +1,337 @@
...
+            webFrontend: `app.test.compactconnect.org`,
+            dataApi: `api.test.compactconnect.org`,
+            s3Upload: `test-persistentstack-mockbulkuploadsbucket0e8f27eb-ivslg22f2yfz.s3.amazonaws.com`,
+        },
+    },
+    ia: {
+        prod: {
+            webFrontend: `app.jcc.iaapi.io`,
+            dataApi: `api.jcc.iaapi.io`,
+            s3Upload: ``, // @TODO: Waiting for environment configuration.
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:32]: Same here.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 15]
```diff
@@ -1,5 +1,337 @@
...
+// ============================================================================
+//                                   HELPERS                                  =
+// ============================================================================
+/**
+ * Get the request domain from the lambda event record.
+ * @param  {object} eventRecord The cloudfront record from the lambda event.
+ * @return {string}             The bare domain of the request domain.
+ */
+const getRequestDomain = (eventRecord) => {
+    let requestDomain = environments.csg.prod.webFrontend;
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:56]: Why is `prod` the default?

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 16]
```diff
@@ -1,5 +1,337 @@
...
+
+// ============================================================================
+//                                   HELPERS                                  =
+// ============================================================================
+/**
+ * Get the request domain from the lambda event record.
+ * @param  {object} eventRecord The cloudfront record from the lambda event.
+ * @return {string}             The bare domain of the request domain.
+ */
+const getRequestDomain = (eventRecord) => {
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:55]: Is it your style guide's standard to use arrow functions here? (What style standards, if any, are you using for JS?)

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 17]
```diff
@@ -1,5 +1,337 @@
...
+/**
+ * Helper to escape CSP src keywords.
+ * @param  {string} keyword The standard keyword value.
+ * @return {string}         The escaped keyword value.
+ */
+const srcKeywordEscape = (keyword) => {
+    let escaped = '';
+
+    if (keyword && typeof keyword === 'string') {
+        escaped = `'${keyword}'`.toLowerCase();
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:132]: how is lowercasing it "escaping" it?

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 18]
```diff
@@ -1,5 +1,337 @@
...
+ */
+const srcKeywordsEscape = (srcList, listName = '') => {
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#keyword_values
+    const srcKeywordsConfig = [
+        { value: 'self', isAllowed: true },
+        { value: 'none', isAllowed: true },
+        { value: 'strict-dynamic', isAllowed: true },
+        { value: 'report-sample', isAllowed: true },
+        { value: 'inline-speculation-rules', isAllowed: true },
+        { value: 'unsafe-inline', isAllowed: true },
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:151]: I see the word `unsafe`, and I get scared 😆 ... do we need to allow this? if yes, add a brief comment above?

File: backend/compact-connect/lambdas/cloudfront-csp/README.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
+# Cloudfront CSP Header Edge Lambda
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: I admittedly don't know AWS well enough, but it seems odd to me that setting CSP headers is in its own lambda. Is that just "the way"?

File: backend/compact-connect/lambdas/cloudfront-csp/README.md [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+_[back to top](#cloudfront-csp-header-edge-lambda)_
+
+---
+## Tests
+- `yarn test`
+
+#### Test output verbosity
+- `index.test.js`
+    - Test block `it()`
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/README.md:45]: What does this mean?

File: .github/workflows/webroot-deploy-development.yml [hunk 21]
```diff
@@ -7,7 +7,7 @@ on:
   # Triggers the workflow on pushes to trunk branches involving changes to web frontend files
```
Comment (jlkravitz) [inline .github/workflows/webroot-deploy-development.yml:7]: What's the reason for the changes in this file?

File: .github/workflows/check-webroot.yml [hunk 22]
```diff
@@ -9,6 +9,7 @@ on:
     branches:
```
Comment (jlkravitz) [inline .github/workflows/check-webroot.yml:9]: What's the reason for the changes in this file?

**Review by jlkravitz — COMMENTED** [review]
Generally looks good. Several questions to help me better understand!

**Update:** Seems like the `unsafe-eval` is needed for i18n? I get an error on deploy that says:

```text
EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
```

File: backend/compact-connect/tests/unit/test_app.py (hunk 1 above)
Comment (jusdino) [reply backend/compact-connect/tests/unit/test_app.py:98]: This change is not related to this PR, but it is a change that should have already been merged in as part of the domain name PR. We had a bit of a knot tied in our git history as we converted from the per-sprint PRs to per-feature PRs. I'm guessing I somehow backed this change back out of my PR when we were untying that knot. Can we keep this in here so it gets corrected with this merge?

File: webroot/vue.config.js (hunk 2 above)
Comment (jsandoval81) [reply webroot/vue.config.js:309]: In this case, it's Mocha + Vue Test Utils. Usually this isn't an issue because libraries either don't have a separate runtime-only version, or if they do then it's trivially handled by Webpack. The vue-i18n library is a one-off in that it needs really specific build / plugin handling. They claim to have addressed this in their latest Beta version, but we're not ready to move to that yet.

File: webroot/src/models/State/State.model.ts (hunk 3 above)
Comment (jsandoval81) [reply webroot/src/models/State/State.model.ts:50]: Good question. It's related to the previous item about the separate runtime-only version of vue-i18n. Their current **non**-runtime-only version has an issue with CSP; and their build bundling is a little chaotic. It was a decision to stick with that library for now, as they claim to have addressed these things in their latest Beta version. Once they've gone stable with that new version, we can evaluate if we want to stick with it, or migrate away.

File: backend/compact-connect/lambdas/cloudfront-csp/test/config/index.js (hunk 4 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/config/index.js:12]: Basically just that CommonJS is still the default in Node; and maybe also that many libraries support CommonJS. Things like ESM tree-shaking aren't ever in play in our backend code. So I guess just simplicity overall.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 5 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:72]: Added a comment in the code.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 6 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:91]: Sometimes SVG images have base64 values / bodies. In those case the `<img src` uses the `data:` scheme. It's oddly common with font & css libraries. Here we're restricting it to image sources.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 7 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js]: Moved to the top of the function.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 9 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js]: Removed.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 10 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js]: Removed.

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 11 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:260]: Factored this to a helper function.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 15 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:56]: Just so if the lambda event structure changes without us knowing, the Prod CSP headers will be returned; keeping the Prod site functional while we work on an update.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 16 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:55]: Yes, we're favoring arrow functions over the `function` keyword (& its associated differences). Our linter styles are using the AirBnB rules as a base, with a small number of modifications.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 17 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:132]: The quotes around `${keyword}` are doing the escaping. The lowercasing is just keeping things uniform.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 18 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js]: Good catch. We weren't actually using it, but it should be disallowed nonetheless.

File: backend/compact-connect/lambdas/cloudfront-csp/README.md (hunk 19 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md]: Good question. This is technically running in the context of a CDN viewer response. I updated this to reflect that.

File: backend/compact-connect/lambdas/cloudfront-csp/README.md (hunk 20 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md]: Yeah, that was a little light on detail 🙂. I updated it to add detail.

File: .github/workflows/webroot-deploy-development.yml (hunk 21 above)
Comment (jsandoval81) [reply .github/workflows/webroot-deploy-development.yml:7]: The domains were recently made permanent, allowing us to stably implement CSP. Updating the domains in this deployment pipeline was part of making it all work together.

File: .github/workflows/check-webroot.yml (hunk 22 above)
Comment (jsandoval81) [reply .github/workflows/check-webroot.yml:9]: Same as in `webroot-deploy-development.yml` above.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 13 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:20]: It'll require the step of setting up the Prod environment. Browsing the board, I didn't see something exactly for that. Do you or @jusdino know if that's part of an existing ticket?

File: backend/compact-connect/tests/unit/test_app.py (hunk 1 above)
Comment (jlkravitz) [reply backend/compact-connect/tests/unit/test_app.py:98]: Got it. Sounds good!

File: backend/compact-connect/lambdas/cloudfront-csp/test/config/index.js (hunk 4 above)
Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/test/config/index.js:12]: Got it, thanks!

File: backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js (hunk 11 above)
Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/test/index.test.js:260]: Thanks!

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 15 above)
Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:56]: makes sense!

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 16 above)
Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:55]: Got it!

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 23]
```diff
@@ -1,5 +1,337 @@
...
+ */
+const buildSrcString = (name = '', list = []) => {
+    let src = [];
+    let srcString = '';
+
+    if (Array.isArray(list)) {
+        src = src.concat(list);
+    }
+
+    srcKeywordsEscape(src, name);
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:198]: Didn't catch this last time, but why not just assert (or similar) that list is an array of strings and remove all the checks in `buildSrcString` and `srcKeywordsEscape`? You may have a good reason but am curious because it seems it would simplify the code significantly by removing many edge cases.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js [hunk 24]
```diff
@@ -1,5 +1,337 @@
...
+                'none',
+            ]),
+            buildSrcString('form-action', [
+                'none',
+            ]),
+            buildSrcString('connect-src', [
+                'self',
+                domains.dataApi,
+                domains.s3Upload,
+                'https://cognito-idp.us-east-1.amazonaws.com',
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/index.js:278]: can we hoist this URL up like we did in the test suite?

File: webroot/src/models/State/State.model.ts (hunk 3 above)
Comment (jlkravitz) [reply webroot/src/models/State/State.model.ts:50]: Got it. Can we make a backlog item that references the relevant code, including this code and that in `StateUpload.ts`, to make an update when the library is updated or if we decide to change translation libraries?

**Review by jlkravitz — CHANGES_REQUESTED** [review]
A couple small follow-ups. Close!

File: webroot/src/models/State/State.model.ts (hunk 3 above)
Comment (jsandoval81) [reply webroot/src/models/State/State.model.ts:50]: Sounds good. I created #118 to track this.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 13 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js:20]: Created #119 to track.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 23 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js]: This `buildSrcString()` function had some extra cruft that I removed. Basically remnants of when multiple optional lists could be passed in. No longer necessary. The rest I feel kinda keeps each function durable & able to stand on its own.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 24 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js]: Done.

File: backend/compact-connect/lambdas/cloudfront-csp/index.js (hunk 23 above)
Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/index.js]: Makes sense, thanks for the explanation!

File: backend/compact-connect/lambdas/cloudfront-csp/README.md [hunk 25]
```diff
@@ -0,0 +1,54 @@
+# Cloudfront Viewer Response Edge Lambda
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: One remaining item- 

Seems like the unsafe-eval is needed for i18n? I get an error on deploy that says:

```EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".```

**Review by jlkravitz — CHANGES_REQUESTED** [review]

File: backend/compact-connect/lambdas/cloudfront-csp/README.md (hunk 25 above)
Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: Oh, interesting. I'm not seeing that in our development environment. Are you sure you're deploying the frontend with the latest build from `yarn build`?

Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: Honestly, that is a fair question. I'll be back...

Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: @jsandoval81 Forgot I needed to upload the `dist/` folder to the S3 bucket.

Regardless, the UI loads, but I get the below error:

```
Refused to connect to 'https://7suc1ey74m.execute-api.us-east-1.amazonaws.com/joshua/mock/providers/query' because it violates the following Content Security Policy directive: "connect-src 'self' https://api.compactconnect.org  https://cognito-idp.us-east-1.amazonaws.com".
```

Is this a configuration issue on my end, or is this a case we didn't handle?

Comment (jsandoval81) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: Technically, a config issue in your stack. The CSP is working as expected in this case by refusing to let the frontend connect to your personal server-side (`https://7suc1ey74m.execute-api.us-east-1.amazonaws.com`).

If that's ok for now then it looks like we're good. Otherwise, we'll need to figure out the best way to add your personal environment to the CSP. That would be most straightforward if you had a long-lived domain for the both your frontend & backend.

Comment (jusdino) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: I don't think we decided on a hard line or anything, but I believe we decided that we would handle two cases for UI:
1) Create a domain name and we support the hosted UI infrastructure (and the CSP works)
2) Run the UI from localhost, point to the API in the cloud (CSP is irrelevant)

Comment (jlkravitz) [reply backend/compact-connect/lambdas/cloudfront-csp/README.md:1]: Hmm OK. This is important though not an immediate priority but, eventually, it should be possible to stand up a test site in a few minutes with at most a few commands and basic technical knowledge. I'm concerned that this is another barrier to entry for doing so.

I plan on sharing the steps I had to take to stand up the site, which I may convert into a PR on the repo's README for your review. Let's continue this discussion at that time and, for now, operate under the status quo described by @jusdino (create a domain name, or run from localhost).

**Review by jlkravitz — APPROVED** [review]
Approving this, but with a note to circle back on https://github.com/csg-org/CompactConnect/pull/116#discussion_r1704686142 in the medium-term.
